### PR TITLE
Backmerge: Fix Reducer Mutations

### PR DESCRIPTION
**Pull request by andre-meyer**
# Backmerge from Olympia
Fixes Reducer that caused mutations

### CODE REVIEW

File: src/reducers/entities.js [hunk 1]
```diff
@@ -13,24 +13,24 @@ const reducer = handleActions({
...
       set(result, `${entityType}`, entitiesState)
 
       keys(action.payload.entities[entityType]).forEach((entityId) => {
         // preserve old entity (only update if not exists)
-        const entityState = get(state, `${entityType}.${entityId}`, {})
+        const entityState = get(Object.assign({}, state), `${entityType}.${entityId}`, {})
         set(result, `${entityType}.${entityId}`, {
           ...entityState,
-          ...action.payload.entities[entityType][entityId],
+          ...Object.assign({}, action.payload.entities[entityType][entityId]),
```
Comment (apanizo) [inline src/reducers/entities.js]: Why are you using the spread operator before ```Object.assign```?

File: src/reducers/entities.js [hunk 2]
```diff
@@ -13,24 +13,24 @@ const reducer = handleActions({
...
   [receiveEntities]: (state, action) => {
-    const result = { ...state }
+    const result = { ...(Object.assign({}, state)) }
     keys(action.payload.entities).forEach((entityType) => {
       // preserve old entities
-      const entitiesState = get(state, `${entityType}`, {})
+      const entitiesState = { ...(get(state, `${entityType}`, {})) }
       set(result, `${entityType}`, entitiesState)
 
       keys(action.payload.entities[entityType]).forEach((entityId) => {
```
Comment (apanizo) [inline src/reducers/entities.js]: Why you just not build an object of entities and once you have finished iterating over them, you just include that new object to the state? In that way, you would save time and resources because if for any reason the state is too big and has other stuff not related with this "reducer" you are copying it in different places, right?

**Review by apanizo — COMMENTED** [review]
I know this part will be refactored (introducing) immutable, so I have just asked a few question just in case I am missing something, I do not understand the reasons for those changes.

File: src/reducers/entities.js [hunk 3]
```diff
@@ -13,24 +13,24 @@ const reducer = handleActions({
...
       },
     },
   }),
   [receiveEntities]: (state, action) => {
-    const result = { ...state }
+    const result = { ...(Object.assign({}, state)) }
     keys(action.payload.entities).forEach((entityType) => {
       // preserve old entities
-      const entitiesState = get(state, `${entityType}`, {})
+      const entitiesState = { ...(get(state, `${entityType}`, {})) }
```
Comment (apanizo) [inline src/reducers/entities.js]: Why you need ```... ``` before the ```(get...)```?

Comment (andre-meyer) [reply src/reducers/entities.js]: `get` returns a reference, this needs to be spread I think:
image

File: src/reducers/entities.js (hunk 1 above)
Comment (mmvsha) [reply src/reducers/entities.js]: Yeah, I've the same question. I think we can get rid of it here. Or maybe it will be fixed after the backmerge is done?

Comment (mmvsha) [reply src/reducers/entities.js]:
```
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };
```
This is a polyfill method from babel which is used to transpile spread operator. So you are doing `Object.assign({}, Object.assign({}, action.payload.entities[entityType][entityId]))`

Comment (andre-meyer) [reply src/reducers/entities.js]: Can you explain to me how I can use `Object.assign` without spreading it? By putting everything into a variable first and then doing Object.assign on it?

Comment (apanizo) [reply src/reducers/entities.js]: https://redux.js.org/docs/recipes/UsingObjectSpreadOperator.html

If it is not clear enough, let me know and I will provide you more concrete examples.

File: src/reducers/entities.js (hunk 3 above)
Comment (apanizo) [reply src/reducers/entities.js]: Normally I am used to see, new data coming actions, and update something in the store at once using ```...``` or ```spread operator``` in this case it seems you need old data from the state as well, I suggest follow the same pattern. Fetch/query data (using lodash), and then at the end build the new state using: ```...oldState, dataFromAction, ...dataFetchedFromState```

Do you understand what I mean?

**Comment by andre-meyer** [comment]
I refactored this (briefly) implementing tests and using `cloneDeep` from lodash to ensure there's no more mutations

**Review by mmv08 — APPROVED** [review]

File: src/reducers/entities.spec.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
+import entityReducer from './entities'
+import { updateEntity, receiveEntities } from 'actions/entities'
+
+describe('entityReducer', () => {
```
Comment (apanizo) [inline src/reducers/entities.spec.js]: Ideally, tests should make easier understand what the code does, where the code is used, and what are corner cases of those use cases.

For doing that, we need to make them more readable. Somethin like should work:
```
// entities.spec.js  <- Good name. it specifies what we are testing
describe('Reducer Entities: Adding users')  <-- Should describe a feature of the domain we are testing
    test("Adding users when the store is empty")   <-- Should specify all different cases
        ...
    test("Adding users different users")
        ...
    test("Modifying existing users")  <-- This test is an example of the typical test added later in time
                                                                  because we have had a bug, like mutations
        //GIVEN
             const storeWithUsers  //names are just an example
             const dataFromEndpointWithSameUsers  //names are just an example
        //WHEN
             //adding to the store              
        //THEN
            //We do not have mutations 
```
If you follow this pattern, you will have better tests, more organised and new developers will understand everything much better and faster.

Comment (andre-meyer) [reply src/reducers/entities.spec.js]: Thanks for the feedback, can you maybe give me a concrete example of how this should look for this reducer?

I rewrote it like this, wdyt?
```jsx
describe('Reducer Entities: Adding and updating Entities', () => {
  describe('Action: updateEntity', () => {
    // GIVEN
    const stateWithTestEntity = {
      testEntity: {
        'test-2': {
          id: 'test-2',
          deeplyNestedData: {
            deeplyNestedKey: 'moreNestedData',
          },
        },
      },
    }

    // WHEN
    const updateAction = updateEntity({ entityType: 'testEntity', data: { id: 'test-1', testObject: { deeplyNested: 'aNestedVaue' } } })
    const stateWithUpdatedEntity = entityReducer(stateWithTestEntity, updateAction)

    // THEN
    test('Updating an Entity', () => {
      // Ensure Action Payload matches Entity Payload
      expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).toMatchObject(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
      expect(updateAction.payload.data.testObject).toMatchObject(stateWithUpdatedEntity.testEntity['test-1'].testObject)
      
      // Ensure no mutations! (=== Compare)
      expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).not.toBe(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
      expect(updateAction.payload.data.testObject).not.toBe(stateWithUpdatedEntity.testEntity['test-1'].testObject)

      // Ensure new Entity was added
      const stateWithNewEntity = {
        'test-2': {
          id: 'test-2',
          deeplyNestedData: {
            deeplyNestedKey: 'moreNestedData',
          },
        },
        'test-1': {
          id: 'test-1',
          testObject: { deeplyNested: 'aNestedVaue' },
        },
      }
      expect(stateWithUpdatedEntity.testEntity).toMatchObject(stateWithNewEntity)
    })
  })
```

Comment (apanizo) [reply src/reducers/entities.spec.js]: Ok, let me ask first a couple of questions:
+  Why do you think it is a good idea to have chained ```describe```? 
+  Which entities are you testing? Users? Trades? Markets?

Comment (andre-meyer) [reply src/reducers/entities.spec.js]: - I'm testing a reducer and inside this reducer, the two specified actions it can receive
- I'm testing the entity reducer, so I'm using an entity called "TestEntity"

Comment (andre-meyer) [reply src/reducers/entities.spec.js]: @apanizo 
Okay, so does this make more sense then?
```javascript
describe('Reducer Entities: Adding and updating Entities', () => {
  test('Updating an Entity', () => {
    // GIVEN
    const stateWithTestEntity = {
      testEntity: {
        'test-2': {
          id: 'test-2',
          deeplyNestedData: {
            deeplyNestedKey: 'moreNestedData',
          },
        },
      },
    }

    // WHEN
    const updateAction = updateEntity({ entityType: 'testEntity', data: { id: 'test-1', testObject: { deeplyNested: 'aNestedVaue' } } })
    const stateWithUpdatedEntity = entityReducer(stateWithTestEntity, updateAction)

    // THEN
    // Ensure Action Payload matches Entity Payload
    expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).toMatchObject(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
    expect(updateAction.payload.data.testObject).toMatchObject(stateWithUpdatedEntity.testEntity['test-1'].testObject)
    
    // Ensure no mutations! (=== Compare)
    expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).not.toBe(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
    expect(updateAction.payload.data.testObject).not.toBe(stateWithUpdatedEntity.testEntity['test-1'].testObject)

    // Ensure new Entity was added
    const stateWithNewEntity = {
      'test-2': {
        id: 'test-2',
        deeplyNestedData: {
          deeplyNestedKey: 'moreNestedData',
        },
      },
      'test-1': {
        id: 'test-1',
        testObject: { deeplyNested: 'aNestedVaue' },
      },
    }
    expect(stateWithUpdatedEntity.testEntity).toMatchObject(stateWithNewEntity)
  })
```

Comment (apanizo) [reply src/reducers/entities.spec.js]: Testing should be done based on features. We can say a feature is something that the user can do in the application.

Orienting tests in that way will help to understand what the code should do and what we expect from that code.  So Testing a reducer is not something a user can do.

So the question is... what the user can do for triggering the usage of that reducer? Maybe load markets and therefore "Add Markets" to the store? Maybe delete a market? 

With that information, I will compose the "describe" functions. For example:
```
describe('Add Markets to store')
describe('Remove markets to store')
```
Once I have the features _described_ the next step is to write the different tests cases. For example:
Which different options can we have when we add markets?
  * Adding markets when the store is empty ```test('Adding markets when store is empty`)```
  * Adding different markets that the ones who are inside the store
  * Finally modify existing markets

The last point is (or the last test case) is the typical test you do not add at the beginning, is the typical test you add when you have had a bug (like mutations).

The last step is how to organize the test. For reading purposes I suggest to use:
```
GIVEN
WHEN
THEN
```
approach.

For example:
```
describe('Add Markets') { 
    test('Add Markets when store is empty') {
        //GIVEN
        const emptyStore = ...;
        const payload = request.payload;

        //WHEN... adding markets to store
        const store = {...emptyStore, payload.markets }

       //THEN... we have the markets in the store
        const markets = getMarketSelector()
       testEqual(market[0], payload.market[0]);
       testEqual(market[1], payload.market[1]);
       ....
    }
    ....
}
describe(' ...') {
    ....
}
.....
```
So I do not think having chained is good in terms of understanding correctly what the app does.

File: src/reducers/entities.spec.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    expect(market.event).toMatchObject(TEST_EVENT)
+
+    expect(market.oracle).not.toBe(TEST_ORACLE) // no mutation check
+    expect(market.eventDescription).not.toBe(TEST_EVENT_DESCRIPTION) // no mutation check
+    expect(market.event).not.toBe(TEST_EVENT) // no mutation check
+  })
+
+  test('Update an existing market', () => {
+    // GIVEN
+    const entityReducerWithMarket = {
```
Comment (apanizo) [inline src/reducers/entities.spec.js]: I would say: ```currentState```

File: src/reducers/entities.spec.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    const updateMarketAction = updateEntity({
+      entityType: 'oracles',
+      data: {
+        id: TEST_ORACLE.address,
+        creator: '0xf00',
+      },
+    })
+
+    // WHEN
+    const entityReducerWithUpdatedMarket = entityReducer(entityReducerWithMarket, updateMarketAction)
```
Comment (apanizo) [inline src/reducers/entities.spec.js]: ... and ```updatedState```

File: src/reducers/entities.spec.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    // THEN
+    const state = { entities: entityReducerWithUpdatedMarket }
+    const markets = getMarkets(state)
```
Comment (apanizo) [inline src/reducers/entities.spec.js]: When refactoring we need to investigate how hard would inject a "real" redux state.

**Review by apanizo — APPROVED** [review]
Really good job. Now if someone needs to understand what the entities reducer does, if he goes to the test he will understand it is used in markets and will understand we have had mutations problems (which they are already fixed).

Probably I would rename a couple of variables, but it looks good to me. 

Congrats. Good job.
